fedimint-recurringd/src/db.rs

```diff
@@ -0,0 +1,196 @@
+use std::collections::HashMap;
+
+use fedimint_core::config::FederationId;
+use fedimint_core::core::OperationId;
+use fedimint_core::db::{AutocommitError, Database, IDatabaseTransactionOpsCoreTyped};
+use fedimint_core::encoding::{Decodable, Encodable};
+use fedimint_core::secp256k1::rand::thread_rng;
+use fedimint_core::secp256k1::PublicKey;
+use fedimint_core::{impl_db_lookup, impl_db_record};
+use fedimint_ln_client::recurring::{PaymentCodeId, PaymentCodeRootKey, RecurringPaymentProtocol};
+use futures::stream::StreamExt;
+use lightning_invoice::Bolt11Invoice;
+use rand::Rng;
+
+#[derive(
+    Debug,
+    Clone,
+    Copy,
+    Eq,
+    PartialEq,
+    Hash,
+    Ord,
+    PartialOrd,
+    fedimint_core::encoding::Encodable,
+    fedimint_core::encoding::Decodable,
+)]
+pub struct FederationDbPrefix([u8; 16]);
+
+impl FederationDbPrefix {
+    pub fn random() -> FederationDbPrefix {
+        FederationDbPrefix(thread_rng().gen())
+    }
+
+    fn prepend(&self, byte: u8) -> Vec<u8> {
```

Comment:
`[u8; 17]` would not require allocation and is still tiny.

Reply:
It's being used as a DB prefix, at that point it would be converted to a `Vec` anyway. If we want to optimize these cases we could use `smallvec` or `tinyvec`.